daemon/k8s_watcher.go

```diff
@@ -1632,3 +1675,120 @@ func (d *Daemon) deleteK8sPodV1(pod *v1.Pod) {
 	}
 
 }
+
+func (d *Daemon) updateK8sNodeTunneling(k8sNodeOld, k8sNodeNew *v1.Node) error {
+	logger := log.WithFields(logrus.Fields{
+		"K8sNodeName": k8sNodeNew.ObjectMeta.Name,
+	})
+
+	nodeNew := k8s.ParseNode(k8sNodeNew)
+	// Ignore own node
+	if nodeNew.Name == node.GetName() {
+		return nil
+	}
+
+	getIDs := func(node *node.Node) (*identity.IPIdentityPair, string, error) {
+		if node == nil {
+			return nil, "", nil
+		}
+		hostIP := node.GetNodeIP(false)
+		if ip4 := hostIP.To4(); ip4 == nil {
+			logger.Debugf("Skipping ipcache update, HostIP is not an IPv4 address %s", hostIP)
+			return nil, "", nil
+		}
+
+		ciliumIPStr := k8sNodeNew.GetAnnotations()[annotation.CiliumHostIP]
+		ciliumIP := net.ParseIP(ciliumIPStr)
+		if ciliumIP == nil {
+			return nil, "", fmt.Errorf("no/invalid Cilium-Host IP: %s", ciliumIPStr)
+		}
+
+		id := &identity.IPIdentityPair{
+			IP:     ciliumIP,
+			HostIP: hostIP,
+			ID:     identity.ReservedIdentityCluster,
+		}
+		return id, ciliumIPStr, nil
+	}
+
+	identityPairNew, ciliumIPStrNew, err := getIDs(nodeNew)
+	if err != nil {
+		return err
+	}
+
+	var identityPairOld *identity.IPIdentityPair
+	if k8sNodeOld != nil {
+		nodeOld := k8s.ParseNode(k8sNodeOld)
+		var (
+			ciliumIPStrOld string
+			err            error
+		)
+		identityPairOld, ciliumIPStrOld, err = getIDs(nodeOld)
+		if err != nil {
+			return err
+		}
+
+		if identityPairOld != nil {
+			if identityPairOld.IP.Equal(identityPairNew.IP) &&
+				identityPairOld.HostIP.Equal(identityPairNew.HostIP) {
+				// If both identities are the same, there's no point on
+				// continuing doing any operation.
+				return nil
+			}
+
+			// Delete previous cilium IP for the given node.
+			ipcache.IPIdentityCache.Delete(ciliumIPStrOld)
```

Comment:
AFAICT `IPCache.Upsert()` should handle this case for you.

Reply:
@joestringer is right. The call to `Upsert()` on line 1744 will handle both insertions and updates. It will delete and re-add the entry if needed and also skip if the mapping is already identical. I think you can remove all code to detect updates.